File: route1-service/src/main/java/pl/edu/pg/benchmarking/route1/RouteService.java
package pl.edu.pg.benchmarking.route1;

import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;

@Service
public class RouteService {
    private final DistanceRepository distanceRepository;
    private final PlaceRepository placeRepository;

    public RouteService(DistanceRepository distanceRepository, PlaceRepository placeRepository) {
        this.distanceRepository = distanceRepository;
        this.placeRepository = placeRepository;
    }

    private Double findDistance(Integer placeId1, Integer placeId2) {
        String coordinates1 = placeRepository.findPlaceCoordinates(placeId1);
        String coordinates2 = placeRepository.findPlaceCoordinates(placeId2);
        return distanceRepository.findDistance(coordinates1, coordinates2);
    }

    private Integer findClosestNeighbor(Integer placeStart, Collection<Integer> places) {
        Integer closestNeighborId = null;
        Double closestNeighborDistance = null;

        for (Integer placeId : places) {
            if (closestNeighborId == null) {
                closestNeighborId = placeId;
                closestNeighborDistance = findDistance(placeStart, placeId);

            } else {
                Double distance = findDistance(placeStart, placeId);
                if (distance < closestNeighborDistance) {
                    closestNeighborId = placeId;
                    closestNeighborDistance = distance;
                }
            }
        }
        return closestNeighborId;
    }

    public List<Integer> computeRoute(Collection<Integer> places) {
        Integer currentPlace = places.stream().findFirst().get();
        int placesAmount = places.size();
        List<Integer> visitedPlaces = new LinkedList<>();
        visitedPlaces.add(currentPlace);
        places.remove(currentPlace);

        for(int i = 1; i < placesAmount; i++) {
            currentPlace = findClosestNeighbor(currentPlace, places);
            visitedPlaces.add(currentPlace);
            places.remove(currentPlace);
        }
        return visitedPlaces;
    }

}

File: .NET/Microservices/Route1Service/Services/RouteService.cs
namespace Route1Service.Services
{
    public class RouteService
    {
        private readonly DistanceRepository _distanceRepository;
        private readonly PlaceRepository _placeRepository;

        public RouteService(DistanceRepository distanceRepository, PlaceRepository placeRepository)
        {
            _distanceRepository = distanceRepository;
            _placeRepository = placeRepository;
        }

        public async Task<double> FindDistanceAsync(int place1Id, int place2Id)
        {
            var cords1 = _placeRepository.FindPlaceCoordinatesAsync(place1Id);
            var cords2 = _placeRepository.FindPlaceCoordinatesAsync(place2Id);

            return await _distanceRepository.FindDistanceAsync(await cords1, await cords2);
        }

        private async Task<int> FindClosestNeighborAsync(int placeStart, List<int> places)
        {
            int? closesNeighborId = null;
            double? closestNeighborDistance = null;

            foreach (var placeId in places)
            {
                if (closesNeighborId == null)
                {
                    closesNeighborId = placeId;
                    closestNeighborDistance = await FindDistanceAsync(placeStart, placeId);
                }

                else
                {
                    var distance = await FindDistanceAsync(placeStart, placeId);
                    if (distance < closestNeighborDistance)
                    {
                        closesNeighborId = placeId;
                        closestNeighborDistance = distance;
                    }
                }
            }

            return closesNeighborId.Value;
        }

        public async Task<List<int>> ComputeRoute(List<int> places)
        {
            var currentPlace = places[0];
            var placeAmount = places.Count;

            var visitedPlaces = new List<int>(places.Count);
            visitedPlaces.Add(currentPlace);
            places.Remove(currentPlace);

            for (var i = 1; i < placeAmount; i++)
            {
                currentPlace = await FindClosestNeighborAsync(currentPlace, places);
                visitedPlaces.Add(currentPlace);
                places.Remove(currentPlace);
            }

            return visitedPlaces;
        }
    }
}
